### PR TITLE
Add support for wildcard versions in pkg.installed states

### LINKED ISSUE #34438: pkg.install's version doesn't work the same way as apt-get's one which is misleading
### Description of Issue/Question

I have `prereq` state which should not be run if there is no changes in another state. Unfortunately I've discovered that it doesn't work as expected due to following message:

```
[DEBUG   ] Current version (['1.11.2-0~wily']) did not match desired version specification (1.11.2*), adding to installation targets
```

meanwhile, `apt-get` does support wildcards:

```
apt-get install docker-engine=1.10.2*
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Selected version '1.10.2-0~wily' (Docker APT Repository:ubuntu-wily [amd64]) for 'docker-engine'
```

I've checked the code, and it seems wildcards are indeed not supported. For instance here I can see that pkg refresh is always called with wildcards due to:
https://github.com/saltstack/salt/blob/develop/salt/modules/aptpkg.py#L502
I propose to use `fnmatch.fnmatchcase` instead of just "==" which will have backward compatibility of course, but will have `apt-get` compliance. 

You might think that wildcards are not necessary, but that is not always convenient to use full package version. Lets look at `docker-engine` package. They have wily postfix if you use `wily` repository, and xenial if you are using `xenial` repository. But version can be the same. Having bunch of if/else makes state file messy. 

I believe we need to introduce 4th argument = compare function here:
https://github.com/saltstack/salt/blob/develop/salt/modules/aptpkg.py#L635
with proper wildcard match or debian package specific normalizing, because `distutils.version.LooseVersion` doesn't do the job it seems. 

P.S. 
While I'm writing this, I would like to take a chance to comment on another odd thing I've discovered while investigating my problem:
https://github.com/saltstack/salt/blob/develop/salt/modules/aptpkg.py#L636
as you can see there is comparison of version ">=" which returns true (mark package for installation?) when desired version is equal or more than current one. Why do we need to mark package for installation if version is equal? Is this expected behaviour?

### PR DESCRIPTION
**Pull request by terminalmage**
This pull request adds support for aptpkg, yumpkg, and pacman to specify a
wildcard in a version number. While apt supports resolving wildcards
intrinsically, this functionality has been added via the use of the
pkg.list_repo_pkgs function for yumpkg and pacman. Code has also been added to
the pkg.installed state to resolve wildcard versions, to preserve idempotence.

An integration test is included, and documentation has been added to the
state's docstring as well as the Nitrogen release notes.

Resolves #34438
Refs: #34442

### CODE REVIEW

File: salt/modules/yumpkg.py [hunk 1]
```diff
@@ -715,7 +795,10 @@ def _parse_output(output, strict=False):
             )
     # Really old version of yum; does not even have --showduplicates option
     if yum_version and yum_version < _LooseVersion('3.2.13'):
-        cmd_prefix = ['yum', '--quiet', 'list']
+        cmd_prefix = ['yum', '--quiet']
+        if cacheonly:
+            yum_prefix.append('-C')
```
Comment (cachedout) [inline salt/modules/yumpkg.py]: `cmd_prefix`?

**Review by cachedout — CHANGES_REQUESTED** [review]
I think there's possibly a small typo here. Also, we need some lint cleaned up.

File: salt/modules/yumpkg.py (hunk 1 above)
Comment (terminalmage) [inline salt/modules/yumpkg.py]: Yeah, typo. Fixed it, as well as lint, and rebased the changes into their respective commits.

**Review by cachedout — APPROVED** [review]

**Comment by cachedout** [comment]
Go Go Jenkins

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by terminalmage** [comment]
Go Go Jenkins!

**Pull request merged**